roachtest: costfuzz failed

roachtest.costfuzz [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=5109241&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=5109241&tab=artifacts#/costfuzz) on master @ [587302906426907122deae6eae5f68630d57e900](https://github.com/cockroachdb/cockroach/commits/587302906426907122deae6eae5f68630d57e900):


```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /artifacts/costfuzz/run_1
	costfuzz.go:181,costfuzz.go:74,test_runner.go:876: error while running perturbed statement: pq: st_dwithin(): error parsing GeoJSON: geojson: unsupported type: 
		(1) attached stack trace
		  -- stack trace:
		  | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.runCostFuzzQuery
		  | 	github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/costfuzz.go:228
		  | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.runOneRoundCostFuzz
		  | 	github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/costfuzz.go:180
		  | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.runCostFuzz
		  | 	github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/costfuzz.go:74
		  | main.(*testRunner).runTest.func2
		  | 	main/pkg/cmd/roachtest/test_runner.go:876
		  | runtime.goexit
		  | 	GOROOT/src/runtime/asm_amd64.s:1581
		Wraps: (2) error while running perturbed statement
		Wraps: (3) pq: st_dwithin(): error parsing GeoJSON: geojson: unsupported type: 
		Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *pq.Error
```
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*costfuzz.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

roachtest.costfuzz [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=5120620&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=5120620&tab=artifacts#/costfuzz) on master @ [98bdf3241028c9b1bdff429fb455e61870adc9d0](https://github.com/cockroachdb/cockroach/commits/98bdf3241028c9b1bdff429fb455e61870adc9d0):


```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /artifacts/costfuzz/run_1
	costfuzz.go:181,costfuzz.go:74,test_runner.go:876: error while running perturbed statement: pq: st_dwithinexclusive(): error parsing EWKT: syntax error: invalid keyword at line 1, pos 0
		(1) attached stack trace
		  -- stack trace:
		  | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.runCostFuzzQuery
		  | 	github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/costfuzz.go:228
		  | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.runOneRoundCostFuzz
		  | 	github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/costfuzz.go:180
		  | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.runCostFuzz
		  | 	github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/costfuzz.go:74
		  | main.(*testRunner).runTest.func2
		  | 	main/pkg/cmd/roachtest/test_runner.go:876
		  | runtime.goexit
		  | 	GOROOT/src/runtime/asm_amd64.s:1581
		Wraps: (2) error while running perturbed statement
		Wraps: (3) pq: st_dwithinexclusive(): error parsing EWKT: syntax error: invalid keyword at line 1, pos 0
		  | LINE 1: X
		  |         ^
		Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *pq.Error
```
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*costfuzz.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

roachtest.costfuzz [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=5129025&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=5129025&tab=artifacts#/costfuzz) on master @ [a86e6c9c4b82cb404e4a20fa70092823fd4a9439](https://github.com/cockroachdb/cockroach/commits/a86e6c9c4b82cb404e4a20fa70092823fd4a9439):


```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /artifacts/costfuzz/run_1
	costfuzz.go:181,costfuzz.go:74,test_runner.go:876: error while running perturbed statement: pq: st_touches(): geos error: TopologyException: side location conflict at -8286154306.8289909 -3240619765.878129 -1450480311.4788685
		(1) attached stack trace
		  -- stack trace:
		  | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.runCostFuzzQuery
		  | 	github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/costfuzz.go:229
		  | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.runOneRoundCostFuzz
		  | 	github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/costfuzz.go:180
		  | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.runCostFuzz
		  | 	github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/costfuzz.go:74
		  | main.(*testRunner).runTest.func2
		  | 	main/pkg/cmd/roachtest/test_runner.go:876
		  | runtime.goexit
		  | 	GOROOT/src/runtime/asm_amd64.s:1581
		Wraps: (2) error while running perturbed statement
		Wraps: (3) pq: st_touches(): geos error: TopologyException: side location conflict at -8286154306.8289909 -3240619765.878129 -1450480311.4788685
		Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *pq.Error
```
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*costfuzz.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

roachtest.costfuzz [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=5134490&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=5134490&tab=artifacts#/costfuzz) on master @ [60d0f95ba057dab8d4f50d2903c504fe275d061a](https://github.com/cockroachdb/cockroach/commits/60d0f95ba057dab8d4f50d2903c504fe275d061a):


```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /artifacts/costfuzz/run_1
	costfuzz.go:181,costfuzz.go:74,test_runner.go:876: error while running perturbed statement: pq: st_pointfromtext(): error parsing EWKT: syntax error: invalid keyword at line 1, pos 0
		(1) attached stack trace
		  -- stack trace:
		  | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.runCostFuzzQuery
		  | 	github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/costfuzz.go:229
		  | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.runOneRoundCostFuzz
		  | 	github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/costfuzz.go:180
		  | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.runCostFuzz
		  | 	github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/costfuzz.go:74
		  | main.(*testRunner).runTest.func2
		  | 	main/pkg/cmd/roachtest/test_runner.go:876
		  | runtime.goexit
		  | 	GOROOT/src/runtime/asm_amd64.s:1581
		Wraps: (2) error while running perturbed statement
		Wraps: (3) pq: st_pointfromtext(): error parsing EWKT: syntax error: invalid keyword at line 1, pos 0
		  | LINE 1: Z
		  |         ^
		Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *pq.Error
```
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*costfuzz.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

roachtest.costfuzz [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=5139612&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=5139612&tab=artifacts#/costfuzz) on master @ [1a91d7cb7b93dfef5dcaf872125875cefa3e0190](https://github.com/cockroachdb/cockroach/commits/1a91d7cb7b93dfef5dcaf872125875cefa3e0190):


```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /artifacts/costfuzz/run_1
	costfuzz.go:181,costfuzz.go:74,test_runner.go:876: error while running perturbed statement: pq: st_intersects(): error parsing EWKT: syntax error: invalid character at line 1, pos 0
		(1) attached stack trace
		  -- stack trace:
		  | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.runCostFuzzQuery
		  | 	github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/costfuzz.go:229
		  | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.runOneRoundCostFuzz
		  | 	github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/costfuzz.go:180
		  | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.runCostFuzz
		  | 	github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/costfuzz.go:74
		  | main.(*testRunner).runTest.func2
		  | 	main/pkg/cmd/roachtest/test_runner.go:876
		  | runtime.goexit
		  | 	GOROOT/src/runtime/asm_amd64.s:1581
		Wraps: (2) error while running perturbed statement
		Wraps: (3) pq: st_intersects(): error parsing EWKT: syntax error: invalid character at line 1, pos 0
		  | LINE 1: SHFX"
		  |         ^
		Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *pq.Error
```
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*costfuzz.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

roachtest.costfuzz [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=5143117&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=5143117&tab=artifacts#/costfuzz) on master @ [7f3c06f5f2c26bc84705430a3622f92ec1444e9d](https://github.com/cockroachdb/cockroach/commits/7f3c06f5f2c26bc84705430a3622f92ec1444e9d):


```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /artifacts/costfuzz/run_1
	costfuzz.go:181,costfuzz.go:74,test_runner.go:876: error while running perturbed statement: pq: st_geographyfromtext(): error parsing EWKT: syntax error: unexpected NUM at line 1, pos 0
		(1) attached stack trace
		  -- stack trace:
		  | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.runCostFuzzQuery
		  | 	github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/costfuzz.go:229
		  | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.runOneRoundCostFuzz
		  | 	github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/costfuzz.go:180
		  | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.runCostFuzz
		  | 	github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/costfuzz.go:74
		  | main.(*testRunner).runTest.func2
		  | 	main/pkg/cmd/roachtest/test_runner.go:876
		  | runtime.goexit
		  | 	GOROOT/src/runtime/asm_amd64.s:1581
		Wraps: (2) error while running perturbed statement
		Wraps: (3) pq: st_geographyfromtext(): error parsing EWKT: syntax error: unexpected NUM at line 1, pos 0
		  | LINE 1: 1R["
		  |         ^
		Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *pq.Error
```
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*costfuzz.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

roachtest.costfuzz [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=5148992&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=5148992&tab=artifacts#/costfuzz) on master @ [bc1ee7c7c276984fce8ff5ba4fcfcdff335dde50](https://github.com/cockroachdb/cockroach/commits/bc1ee7c7c276984fce8ff5ba4fcfcdff335dde50):


```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /artifacts/costfuzz/run_1
	costfuzz.go:181,costfuzz.go:74,test_runner.go:876: error while running perturbed statement: pq: st_box2dfromgeohash(): invalid GeoHash: geohash decode 'npfdz': invalid character at index 2
		(1) attached stack trace
		  -- stack trace:
		  | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.runCostFuzzQuery
		  | 	github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/costfuzz.go:229
		  | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.runOneRoundCostFuzz
		  | 	github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/costfuzz.go:180
		  | github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.runCostFuzz
		  | 	github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/costfuzz.go:74
		  | main.(*testRunner).runTest.func2
		  | 	main/pkg/cmd/roachtest/test_runner.go:876
		  | runtime.goexit
		  | 	GOROOT/src/runtime/asm_amd64.s:1581
		Wraps: (2) error while running perturbed statement
		Wraps: (3) pq: st_box2dfromgeohash(): invalid GeoHash: geohash decode 'npfdz': invalid character at index 2
		Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *pq.Error
```
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*costfuzz.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/cmd/reduce/main.go
+++ b/pkg/cmd/reduce/main.go
@@ -58,6 +58,9 @@ var (
 	tlp             = flags.Bool("tlp", false, "last two non-empty lines in file are equivalent queries returning different results")
 )
 
+// TODO(michae2): Add -costfuzz mode which is aware that the last three
+// statements are two identical queries separated by a setting change.
+
 const description = `
 The reduce utility attempts to simplify SQL that produces an error in
 CockroachDB. The problematic SQL, specified via -file flag, is

--- a/pkg/cmd/roachtest/tests/costfuzz.go
+++ b/pkg/cmd/roachtest/tests/costfuzz.go
@@ -223,13 +223,27 @@ func runCostFuzzQuery(
 	// Then, rerun the statement with cost perturbation.
 	rows2, err := conn.Query(stmt)
 	if err != nil {
-		logStmt(stmt)
-		logStmt(seedStmt)
-		logStmt(stmt)
-		return errors.Wrap(err, "error while running perturbed statement")
+		// If the perturbed plan fails with an internal error while the normal plan
+		// succeeds, we'd like to know, so consider this a test failure.
+		es := err.Error()
+		if strings.Contains(es, "internal error") {
+			logStmt(stmt)
+			logStmt(seedStmt)
+			logStmt(stmt)
+			return errors.Wrap(err, "internal error while running perturbed statement")
+		}
+		// Otherwise, skip perturbed statements that fail with a non-internal
+		// error. This could happen if the statement contains bad arguments to a
+		// function call, for example, and the normal plan was able to skip
+		// evaluation of the function due to short-circuiting (see #81032 for an
+		// example).
+		//nolint:returnerrcheck
+		return nil
 	}
 	defer rows2.Close()
 	perturbedRows, err := sqlutils.RowsToStrMatrix(rows2)
+	// If we've gotten this far, we should be able to print the results of the
+	// perturbed statement, so consider it a test failure if we cannot.
 	if err != nil {
 		logStmt(stmt)
 		logStmt(seedStmt)
@@ -248,6 +262,11 @@ func runCostFuzzQuery(
 			diff, stmt,
 		)
 	}
+
+	// TODO(michae2): If we run into the "-0 flake" described in PR #79551 then
+	// we'll need some other strategy for comparison besides diffing the printed
+	// results. One idea is to CREATE TABLE AS SELECT with both queries, and then
+	// EXCEPT ALL the table contents. But this might be very slow.
 
 	// Finally, disable cost perturbation for the next statement.
 	resetSeedStmt := "RESET testing_optimizer_random_cost_seed"

